src/rules/maxNestingDepthRule.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isFunctionScopeBoundary, isIdentifier } from "tsutils";
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static DEFAULT_THRESHOLD = 4;
+    public static MINIMUM_THRESHOLD = 0;
+
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "max-nesting-depth",
+        description: "Enforces a maximum threshold of nesting depth.",
+        descriptionDetails: Lint.Utils.dedent`
+            The nesting depth of the code blocks is a metric determined for each method.
+            It is computed by determining the nesting levels of the control statements within the method.
+            A starting value of 0 is assigned, then this value is
+            incremented each time a new block is entered, and decremented
+            each time a block is left.
+            The following statements contribute to nesting depth:
+            * \`if\` statements
+            * \`switch\` statements
+            * \`for\`, \`for in\` and \`for of\` loops
+            * \`while\` and \`do while\` loops
+            * \`try\` \`catch\` and \`finally\` blocks`,
+        rationale: Lint.Utils.dedent`
+            Maximum nesting depth is a code metric which indicates the level of imbrication in a
+            function. This metric quickly reveals the methods whose code is made difficult to read
+            by an excessive nesting of the control statements.
+
+            It's better to have smaller, single-purpose functions with self-documenting names.`,
+        optionsDescription: Lint.Utils.dedent`
+            An optional upper limit for maximum nesting depth can be specified. If no limit option
+            is provided a default value of ${Rule.DEFAULT_THRESHOLD} will be used.`,
+        options: {
+            type: "number",
+            minimum: Rule.MINIMUM_THRESHOLD,
+        },
+        optionExamples: [true, [true, 5]],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(expected: number, actual: number, name?: string): string {
+        return (
+            `The function${name === "" ? "" : ` ${name}`} has a nesting depth of at least ` +
```

Comment:
I used the wording "_has a nesting depth of at least N_" since I stop parsing the function as soon as the maximum nesting depth is reached.
So the function nesting depth is _at least_ equal to the maximum nesting depth, but it might be greater.

Reply:
On second thought - I think it could be a little frustrating for users to not know how far away their function's nesting depth is. Not explicitly stating the nesting depth for functions also means it'll be hard for folks to experimentally figure out what syntax types increase it.

Thoughts @giladgray?

Reply:
I chose to stop parsing the AST as soon as possible after reading these hints: https://palantir.github.io/tslint/develop/custom-rules/performance-tips.html
But I could change the implementation to display a message for each statement deeper than the maximum allowed depth (as ESLint does).

Reply:
Happy new year !
Any update @JoshuaKGoldberg?
Any thoughts @giladgray?

Reply:
Happy new year! 🎉

Waiting on the implementation to mention the real nesting depth number.